### ISSUE TITLE
[invokers] loseinterest should include a "safe area" mechanism

### ISSUE THREAD

**Issue by lukewarlow** [openui/open-ui#963]
The concept of a safe triangle (or other shape) between the interest Invoker and the invoked element is often implemented by Devs. This is often used to help meet WCAG (https://www.w3.org/WAI/WCAG22/Understanding/content-on-hover-or-focus.html).

We should ensure that interest target is designed in such a way that this concept is either built in by default or trivially enabled.

**Comment by lukewarlow** [comment]
https://x.com/claviska/status/1729964684441223319?s=20 this is just one example of many that quickly provides a visual demonstration of what I mean

**Comment by lukewarlow** [comment]
https://www.smashingmagazine.com/2023/08/better-context-menus-safe-triangles/ also provides a good explainer

**Comment by brechtDR** [comment]
It certainly is an interesting concept. Not sure of this something that we could achieve with CSS in the future and adding a pseudo element to the tooltip to create the lace. 🤔 Should try this out.

**Comment by lukewarlow** [comment]
Also should consider it this should be linked to CSS anchor positioning somehow or if the positioning mechanism is tangential to this concept.

What happens if the hover triggers something far away from the invoker? Do we set a max size for the safe area before we don't bother?

**Comment by keithamus** [comment]
@mfreed7 originally mentioned this here: https://github.com/openui/open-ui/issues/905#issuecomment-1781167595 and I’m fully on board but we deferred it while the proposal migrated repos and the idea got left on the cutting room floor, so glad you’ve revived it and given it a proper issue.

**Comment by mfreed7** [comment]
I'm also uncertain about the "safe triangle" behavior - could that be spec'd in a clean way for all possible triggers and targets?

I do think that even if we add the "safe triangle" concept, we still need a developer-controllable delay after which the target gets the `loseinterest` event. And it needs to be properly counted as time away from *either* the invoker *or* the target element. I.e. if the delay is set to 1 second, and you move from the invoker to the target in 0.5 seconds, the counter resets to zero. You need to leave them both for 1 second for the `loseinterest` to fire. If the "safe triangle" is added, staying inside the triangle would similarly reset the counter.

**Comment by lukewarlow** [comment]
So thinking on the safe area Vs delay thing a bit more, delaying (or not firing) lose interest doesn't actually achieve much in the case where another interest Invoker is nearby. Because chances are a new popover will show which will dismiss the existing one.

It would need to be a delay on triggering a loseinterest AND the following interest event.

**Comment by mfreed7** [comment]
> So thinking on the safe area Vs delay thing a bit more, delaying (or not firing) lose interest doesn't actually achieve much in the case where another interest Invoker is nearby. Because chances are a new popover will show which will dismiss the existing one.
> 
> It would need to be a delay on triggering a loseinterest AND the following interest event.

Help me understand this comment. The concept I had in mind is taken from my original proposal for popover. There are two delays - one for the `interest` event and another for the `loseinterest` event. But for this example, assume they're both set to `1 second` for all invokers. Further assume we're using a mouse, so "hover" is how we get/lose interest. Ok:

1. Hover invoker <span>#</span>1 for 0.5 seconds. (no `interest` events fired)
2. Hover invoker <span>#</span>2 for 1 second. (invoker <span>#</span>2 fires `interest`)
3. Hover invoker <span>#</span>1 again for 0.5 seconds. (no events fired)
4. Re-hover invoker <span>#</span>2. (no events fired)
5. Wait 2 seconds. (still no events fired)
6. Hover invoker <span>#</span>1 for 1 second (`loseinterest` fired for invoker <span>#</span>2, then `interest` fired for invoker <span>#</span>1)

And as you said, if the delays and actions were set so that an `interest` on a new invoker fires before the `loseinterest` on another popover-invoker, then the popover logic will cause the first popover to be dismissed because the new popover is being shown. But that's separate from this proposal.

I think we're in agreement, I just want to make sure.

**Comment by kbrilla** [comment]
But delay should not be used for loseintrest on keyboard navigation, or It
should be able to configure it separately

W dniu pon., 4.12.2023 o 19:36 Luke Warlow ***@***.***>
napisał(a):
